Switch to Go modules to manage dependencies

cbandy opened the pull request:
**Checklist:**

 <!--- Make sure your PR is documented and tested before submission. Put an `x` in all the boxes that apply: -->
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?

**Type of Changes:**

 <!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
 - [x] Breaking change (fix or feature that would cause existing functionality to change)


**What is the new behavior (if this is a feature change)?**

The `go` command is sufficient to install dependencies, run tests, and build binaries. We no longer depend on the `GOPATH` or `GOBIN` environment variables, and the project can reside outside your GOPATH. 

**Other Information**:

Once you have these changes locally, you'll want to run `make clean-deprecated` or `make clean` to remove build artifacts leftover from the old ways.

Replaces: #1648
Issue: [ch9453]
fixes #1647

andrewlecuyer commented on docs/content/contributing/developer-setup.md:
```diff
@@ -21,14 +21,10 @@ The following environment variables are expected by the steps in this guide:
-`PGO_BASEOS` | {{< param centosBase >}} | Base OS for container images
 `PGO_CMD` | kubectl | Cluster management tool executable
-`PGO_IMAGE_PREFIX` | crunchydata | Container image prefix
 `PGO_OPERATOR_NAMESPACE` | pgo | Kubernetes namespace for the operator
```
Was just curious why we're removing some of these (`PGO_BASEOS`, `PGO_OPERATOR_NAMESPACE` & `PGO_VERSION`) and not the others.  All of these appear to be defined in the default Makefile, so just wondering which we're keeping in this list.

Also, would `PGO_OPERATOR_NAMESPACE` be more Bash install thing vs. dev setup?  I do see we have a `Deployment` section in the instructions that references what is essentially the Bash install, which would require this env var.  Maybe we should just point to the bash install instructions instead (especially considering the various other setup, e.g. for things like RBAC, CRD's, etc, are really needed to properly deploy)?

cbandy replied:
> Was just curious why we're removing some of these (`PGO_BASEOS`, `PGO_OPERATOR_NAMESPACE` & `PGO_VERSION`) and not the others. All of these appear to be defined in the default Makefile, so just wondering which we're keeping in this list.

I intended to remove only the variables related to building images because those are only used in the Makefile. However, I see that `PGO_IMAGE_PREFIX` is also used in `bin/push-to-gcr.sh`.

> Also, would `PGO_OPERATOR_NAMESPACE` be more Bash install thing vs. dev setup? I do see we have a `Deployment` section in the instructions that references what is essentially the Bash install, which would require this env var. Maybe we should just point to the bash install instructions instead (especially considering the various other setup, e.g. for things like RBAC, CRD's, etc, are really needed to properly deploy)?

I'm fine with linking to the Bash install docs for the Deployment section. I really just punted on figuring out when/why `PGO_CMD` and `PGO_OPERATOR_NAMESPACE` should be set because they are used throughout the `deploy` directory.

andrewlecuyer commented on Makefile:
```diff
@@ -73,10 +71,12 @@ ifeq ("$(PGO_BASEOS)", "centos8")
 
 DEBUG_BUILD ?= false
-GCFLAGS=
+GO_BUILD = $(GO_CMD) build
```
Any chance we might still be able to allow for `install` vs `build`?

I currently leverage this in conjunction with my `GO_BIN` to quickly/easily launch our various Operator applications as part of my typical development process, since `GO_BIN` is already part of my default`PATH` (with this change I suppose I could also make `$PGOROOT/bin` part of my standard `PATH`).

cbandy replied:
I don't like `install` because it creates things outside the project directory.

I use `./bin/pgo` to run the client, but (most of) the other binaries go into subdirectories of `./bin` which I can see being a hassle. I'd like to move those up into `./bin` so that `$PGOROOT/bin` can be added to `PATH` as you suggest. Do you know of anything besides the Dockerfiles that would be affected by that?

cbandy replied:
So, most executables are named the same as a non-empty subdirectory in `bin`. I can move `postgres-operator` up alongside `pgo` and `apiserver` at least. Would that alleviate most of the pain?

andrewlecuyer replied:
I am definitely good with that approach.  With `pgo`, `postgres-operator` and `apiserver` being the most common binaries run via my `PATH`, this should definitely alleviate most pain (I am also fine with this approach considering that some of the various other binaries such as `pgo-scheduler`, `pgo-rmdata` & `pgo-backrest` will likely change in some way or another moving forward, e.g. `pgo-backrest` will soon be moved to the Container Suite project as part of compaction).

cbandy replied:
Done.

andrewlecuyer commented on Makefile:
```diff
@@ -94,76 +94,65 @@ images = pgo-apiserver \
 cleannamespaces:
-	cd deploy && ./cleannamespaces.sh
-
-bounce:
-	$(PGO_CMD) \
-		--namespace=$(PGO_OPERATOR_NAMESPACE) \
-		get pod \
-		--selector=name=postgres-operator \
-		-o=jsonpath="{.items[0].metadata.name}" \
-	| xargs $(PGO_CMD) --namespace=$(PGO_OPERATOR_NAMESPACE) delete pod
```
With this section gone, can `PGO_OPERATOR_NAMESPACE` be removed from the Makefile?

cbandy replied:
Yep. Looks like `PGO_CMD` can go, too.

andrewlecuyer commented on bin/get-deps.sh:
```diff
@@ -57,36 +37,21 @@ if ! [ -f $EVTDIR/nsqd -a -f $EVTDIR/nsqadmin ]; then
-	else
-		sudo yum -y install docker --enablerepo=rhel-7-server-extras-rpms
-	fi
+if ! command -v docker &> /dev/null && ! command -v podman &> /dev/null; then
+	echo 'Cannot find `docker` nor `podman`. Perhaps:'
+	echo '  sudo yum install docker'
+	exit 1
```
Since technically neither `docker` nor `podman` are required, should we remove this check?

For instance, if using this script from a CI server, I wouldn't want it to simply exit without the presence of either or these two tools, since I might be using something like `buildah` and `crio` only (and would still want to be able to grab all of the other dependencies, despite this check failing).

cbandy replied:
Yep, will do.

andrewlecuyer reviewed (commented):
Looks great, just a few questions/comments.

cbandy commented:
> 2. I'd be completely fine with eliminating `odev` and its reference entirely.

Done.

I've addressed most comments by amending the last two commits.
